Runtime/Scripts/DataCollection/RGDataCollection.cs

```diff
@@ -32,20 +37,56 @@ namespace RegressionGames.DataCollection
      *
      * Once all clients have been disconnected, this class can be reset and all data left over (i.e. screenshots) can
      * be deleted from the system.
-     * 
+     *
      */
     public class RGDataCollection
     {
-        
+        private static readonly ProfilerMarker LogCollectionMarker = new(ProfilerCategory.Ai, "RegressionGames.DataCollection.LogCollector");
+        private static readonly ProfilerMarker SaveTickDataMarker = new(ProfilerCategory.Ai, "RegressionGames.DataCollection.SaveTickData");
+
         private readonly string _sessionName = Guid.NewGuid().ToString();
-        private readonly Dictionary<long, RGBot> _clientIdToBots = new();
-        private readonly Dictionary<long, DateTime> _clientIdStartTimes = new();
-        private readonly ConcurrentDictionary<long, List<Task>> _clientIdToReplayDataTasks = new();
-        private readonly ConcurrentDictionary<long, Task> _clientIdToValidationDataTask = new();
+        private readonly ConcurrentDictionary<long, BotInstanceDataCollectionState> _clientIdToState = new();
         private readonly string _rootPath = Application.persistentDataPath;
         private readonly ConcurrentQueue<long> _screenshotTicksRequested = new();
         private readonly ConcurrentDictionary<long, long> _screenshotTicksCaptured = new();
-        private readonly ConcurrentDictionary<long, RGValidationSummary> _clientIdToValidationSummaries = new();
+
+        private class BotInstanceDataCollectionState
+        {
+            public long clientId;
+            public RGBot bot;
+            public ConcurrentQueue<LogDataPoint> logs = new();
+            public DateTime startTime;
+            public List<Task> replayDataTasks = new();
+            public Task validationDataTask;
+            public Task logFlushTask;
+            public RGValidationSummary validationSummary = new(0,0,0);
+
+            public BotInstanceDataCollectionState(RGBot bot, DateTime startTime)
+            {
+                this.bot = bot;
+                this.startTime = startTime;
+            }
+
+            public void StartCapturingLogs()
+            {
+                Application.logMessageReceivedThreaded += LogMessageReceived;
+            }
+
+            public void StopCapturingLogs()
+            {
+                Application.logMessageReceivedThreaded -= LogMessageReceived;
+            }
+
+            private void LogMessageReceived(string message, string stacktrace, LogType type)
```

Comment:
Update from the previous iteration here.

Previously, we used a single `logMessageReceivedThreaded` handler and iterated through all the active bots to stash the log message in the queue. I did some profiling and when logging 100 messages per tick, **and** enabling "Deep Tracing" (aka Instrumentation, where each function call is measured and counted), I did see some impact from this. So, I tried a refactoring that I had been considering before profiling and it did yield a _slightly_ noticeable improvement. It's worth noting that I had to really dig deep to find this call in the profile at all, so I'm not sure it'll be a big deal either way, but I went with the "faster" method anyway.

Basically, we add a separate handler for `logMessageReceivedThreaded` for each bot. This means that when a log message arrives, the built-in .NET event dispatching logic runs and calls each handler in sequence. Since this handler holds a reference to the `BotInstanceDataCollectionState` instance on which it is defined, it has immediate access to the `logs` queue, without having to look it up in the lock-heavy ConcurrentDictionary structure. This keeps thread contention to a minimum.

The trade-off is that we have to allocate a `LogDataPoint` for each bot in this approach. I didn't see a measurable impact from that, but we can always optimize this further later. We have a lot more allocations throughout the data collection process that could be optimized away, so there's plenty of fodder for a future optimization pass through here to speed things up more.